### ISSUE TITLE
ihaskell-widgets runs with no error, however, displays no widget

### ISSUE THREAD

**Issue by eldad-a** [gibiansky/IHaskell#870]
Hello,

The short version:
ihaskell-widgets runs with no error, however shows no widget.

Long version:
This is my first experience with `Haskell` and I am thankful to the Devs for creating an interface based on `jupyter notebook`, which I find very friendly.

The following example code executes with no error, however, with no widget displayed either.

```haskell
import IHaskell.Display.Widgets
slider <- mkFloatSlider

setField slider ChangeHandler $ do
    value <- getField slider FloatValue
    print value
    
slider
```
The above is copied from [Demo of IHaskell Notebook](https://begriffs.com/posts/2016-01-20-ihaskell-notebook.html) by [Andew Giabiansky](http://andrew.gibiansky.com/pages/about.html), January 20, 2016

To run the example notebooks from the `ihaskell-widgets` [IHaskell repository](https://github.com/gibiansky/IHaskell) I had to rename the `stack.yaml` in the `ihaskell-display/ihaskell-widgets/`, as suggested in [Cannot install ihaskell-display widgets with stack #715
](https://github.com/gibiansky/IHaskell/issues/715) .

Then the examples (two notebooks I tried) run with no errors, but still, no widget is displayed.

Hope you could guide me how to resolve this matter.

**Comment by vaibhavsagar** [comment]
It looks like `ihaskell-widgets` hasn't been updated in some time (beyond version bumps by me) and the protocol version it speaks is out of date. I'm not sure when I'll get around to fixing this and I'm sorry I don't have a better answer for you!

![ihaskell-widget-protocol-mismatch](https://user-images.githubusercontent.com/1525767/40951723-c51159a0-68aa-11e8-9f12-2b245f531519.png)

**Comment by vaibhavsagar** [comment]
This issue looks related: https://github.com/JuliaGizmos/Interact.jl/issues/176

**Comment by vaibhavsagar** [comment]
For myself or whoever decides to implement this: here is [a document describing the new protocol](https://github.com/jupyter-widgets/ipywidgets/blob/3023b3ed24fbb172eb44643798a603aaed15ae76/packages/schema/messages.md) and the changes would probably have to made to [`IHaskell.Display.Widgets`](https://github.com/gibiansky/IHaskell/blob/00c762b77b36b08bb7baf4a3b84f251bd9b6be90/src/IHaskell/Eval/Widgets.hs).

**Comment by eldad-a** [comment]
@vaibhavsagar 

A.  Many thanks for the quick and detailed reply!

B. 
> [This issue looks related: JuliaGizmos/Interact.jl#176](https://github.com/gibiansky/IHaskell/issues/870#issuecomment-394562129
)

Following that [link](https://github.com/JuliaGizmos/Interact.jl/issues/176) I did:
`pip3 uninstall ipywidgets`
followed by 
`pip install ipywidgets==6.0.0`

and the [slider example above](https://github.com/gibiansky/IHaskell/issues/870#issue-329190009) works :-)

**Comment by vaibhavsagar** [comment]
Great to see you found a workaround!

**Comment by kyagrd** [comment]
Tried the workaround suggested by eladad-a for a Docker image but hasn't been successful.
If anyone managed to work around this problem and built a Docker image I'd be interested to know about it.

**Comment by jamesdbrock** [comment]
*ipywidgets* version 7.0 changelog

https://ipywidgets.readthedocs.io/en/stable/changelog.html#id5

**Comment by jamesdbrock** [comment]
So to summarize and repeat what's already been said here by @vaibhavsagar :

IHaskell Widgets currently supports the [Widget messaging protocol version 1](https://github.com/jupyter-widgets/ipywidgets/blob/3023b3ed24fbb172eb44643798a603aaed15ae76/packages/schema/messages.md#widget-messaging-protocol-version-1) , implemented in ipywidgets up to version 6.0.1.

[ipywidgets from version 7.0.0](https://pypi.org/project/ipywidgets/#history) implements the [Widget messaging protocol version 2](https://github.com/jupyter-widgets/ipywidgets/blob/3023b3ed24fbb172eb44643798a603aaed15ae76/packages/schema/messages.md#widget-messaging-protocol-version-2).

The work would involve these two packages:

1. [IHaskell/Eval/Widgets.hs](https://github.com/gibiansky/IHaskell/blob/master/src/IHaskell/Eval/Widgets.hs)
2. [ihaskell-display/ihaskell-widgets](https://github.com/gibiansky/IHaskell/tree/master/ihaskell-display/ihaskell-widgets)

I would approach this work in these steps:

1. Build a test environment with `ipywidgets==6.0.1` in which `ihaskell-widgets` is working.
2. Upgrade IHaskell to widget messaging protocol 2.
3. Build a test environment with `ipywidgets==7.4.2` in which `ihaskell-widgets` is working.

It would be nice if the test environment could be included in the IHaskell repo.

I would be very grateful if @sumitsahrawat or @gibiansky would offer any advice about this upgrade. Especially advice about getting debugging visibility into the comms.

**Comment by jamesdbrock** [comment]
The `stack --docker` configuration from @jcberentsen bc6d9cad46d3e7dada705d80116c734e28885549 is going to be a good way to set up the test environment.

**Comment by rehno-lindeque** [comment]
> Especially advice about getting debugging visibility into the comms.

Not sure if this is what was meant, but just to mention: it's possible to see websocket messages for comms in the chrome browser's development tools. I found that there's an buttons to clear each messages on each individual websocket (don't clear all network traffic; that prevents you from seeing new messages).

**Comment by jamesdbrock** [comment]
I think it's gonna turn out that issue https://github.com/gibiansky/IHaskell/issues/1144 is required to resolve this issue.

**Comment by jamesdbrock** [comment]
It looks like __jupyter-widgets__ is prepping for version 8. https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/schema/jupyterwidgetmodels.v8.md

**Comment by jamesdbrock** [comment]
https://github.com/jupyter-widgets/ipywidgets/issues/2798#issuecomment-592076970

> So the story to users/developers is:
>
> * ipywidgets 7 fully supports jlab 1 and jlab 2. Custom widgets will need a patch release to add support for jlab 2 (just like they did for jlab 1)
> * ipywidgets 8 supports jlab 2 (but not jlab 1). Custom widgets will need a major release to support ipywidgets 8
> * Both ipywidgets 7 and 8 support classic notebook

**Comment by jamesdbrock** [comment]
ipywidgets changelog: https://github.com/jupyter-widgets/ipywidgets/blob/master/docs/source/changelog.md#70

**Comment by jamesdbrock** [comment]
@daviddavo is working on this issue for Google Summer of Code 2021 and he is making excellent progress.

**Comment by jamesdbrock** [comment]
This issue has been solved by #1242 

Thx David!
